File: sdks/ios/ConsentOS/Sources/ConsentOSCore/ConsentState.swift
import Foundation

/// Represents the complete consent state for a visitor.
///
/// This model mirrors the web consent cookie structure for cross-platform consistency.
/// It is persisted locally via ``ConsentStorage`` and synced to the server.
public struct ConsentState: Codable, Equatable, Sendable {

    // MARK: - Properties

    /// A stable, anonymous identifier for this device/visitor.
    /// Generated once and persisted across sessions.
    public let visitorId: String

    /// The set of categories the visitor has explicitly accepted.
    public var accepted: Set<ConsentCategory>

    /// The set of categories the visitor has explicitly rejected.
    public var rejected: Set<ConsentCategory>

    /// The timestamp at which consent was last recorded.
    public var consentedAt: Date?

    /// The banner configuration version active when consent was collected.
    /// Used to detect when consent must be re-collected after a config change.
    public var bannerVersion: String?

    /// Whether the user has interacted with the banner (accepted or rejected).
    ///
    /// Returns `false` when the state represents the pre-consent default.
    public var hasInteracted: Bool {
        consentedAt != nil
    }

    // MARK: - Derived State

    /// Returns `true` if the user has granted consent for the given category.
    ///
    /// `necessary` is always considered granted regardless of the stored state.
    public func isGranted(_ category: ConsentCategory) -> Bool {
        guard category != .necessary else { return true }
        return accepted.contains(category)
    }

    /// Returns `true` if the user has explicitly denied consent for the given category.
    public func isDenied(_ category: ConsentCategory) -> Bool {
        guard category != .necessary else { return false }
        return rejected.contains(category)
    }

    // MARK: - Initialisers

    /// Creates a new, blank consent state for the given visitor.
    ///
    /// No categories are accepted or rejected; `consentedAt` is `nil`.
    public init(visitorId: String) {
        self.visitorId = visitorId
        self.accepted = []
        self.rejected = []
        self.consentedAt = nil
        self.bannerVersion = nil
    }

    /// Creates a fully populated consent state.
    public init(
        visitorId: String,
        accepted: Set<ConsentCategory>,
        rejected: Set<ConsentCategory>,
        consentedAt: Date?,
        bannerVersion: String?
    ) {
        self.visitorId = visitorId
        self.accepted = accepted
        self.rejected = rejected
        self.consentedAt = consentedAt
        self.bannerVersion = bannerVersion
    }

    // MARK: - Mutations

    /// Returns a new state with all non-necessary categories accepted.
    public func acceptingAll() -> ConsentState {
        let allOptional = ConsentCategory.allCases.filter { $0.requiresConsent }
        return ConsentState(
            visitorId: visitorId,
            accepted: Set(allOptional),
            rejected: [],
            consentedAt: Date(),
            bannerVersion: bannerVersion
        )
    }

    /// Returns a new state with all non-necessary categories rejected.
    public func rejectingAll() -> ConsentState {
        let allOptional = ConsentCategory.allCases.filter { $0.requiresConsent }
        return ConsentState(
            visitorId: visitorId,
            accepted: [],
            rejected: Set(allOptional),
            consentedAt: Date(),
            bannerVersion: bannerVersion
        )
    }

    /// Returns a new state accepting only the specified categories (and rejecting the rest).
    public func accepting(categories: Set<ConsentCategory>) -> ConsentState {
        let allOptional = Set(ConsentCategory.allCases.filter { $0.requiresConsent })
        let toAccept = categories.filter { $0.requiresConsent }
        let toReject = allOptional.subtracting(toAccept)
        return ConsentState(
            visitorId: visitorId,
            accepted: toAccept,
            rejected: toReject,
            consentedAt: Date(),
            bannerVersion: bannerVersion
        )
    }
}

File: sdks/android/consentos/src/main/kotlin/com/consentos/ConsentState.kt
package com.consentos

import kotlinx.serialization.Serializable

/**
 * Represents the complete consent state for a visitor.
 *
 * This model mirrors the web consent cookie structure for cross-platform consistency.
 * It is persisted locally via [ConsentStorage] and synced to the server.
 */
@Serializable
data class ConsentState(
    /**
     * A stable, anonymous identifier for this device/visitor.
     * Generated once and persisted across sessions.
     */
    val visitorId: String,

    /** The set of categories the visitor has explicitly accepted. */
    val accepted: Set<ConsentCategory> = emptySet(),

    /** The set of categories the visitor has explicitly rejected. */
    val rejected: Set<ConsentCategory> = emptySet(),

    /**
     * The timestamp (epoch milliseconds) at which consent was last recorded.
     * `null` until the user interacts with the banner.
     */
    val consentedAtMs: Long? = null,

    /**
     * The banner configuration version active when consent was collected.
     * Used to detect when consent must be re-collected after a config change.
     */
    val bannerVersion: String? = null,
) {

    // -------------------------------------------------------------------------
    // Derived State
    // -------------------------------------------------------------------------

    /**
     * Whether the user has interacted with the banner (accepted or rejected).
     *
     * Returns `false` when the state represents the pre-consent default.
     */
    val hasInteracted: Boolean
        get() = consentedAtMs != null

    /**
     * Returns `true` if the user has granted consent for the given [category].
     *
     * [ConsentCategory.NECESSARY] is always considered granted regardless of the stored state.
     */
    fun isGranted(category: ConsentCategory): Boolean {
        if (category == ConsentCategory.NECESSARY) return true
        return category in accepted
    }

    /**
     * Returns `true` if the user has explicitly denied consent for the given [category].
     */
    fun isDenied(category: ConsentCategory): Boolean {
        if (category == ConsentCategory.NECESSARY) return false
        return category in rejected
    }

    // -------------------------------------------------------------------------
    // Mutations — returns new instances (immutable pattern)
    // -------------------------------------------------------------------------

    /**
     * Returns a new state with all non-necessary categories accepted.
     */
    fun acceptingAll(): ConsentState {
        val allOptional = ConsentCategory.entries.filter { it.requiresConsent }.toSet()
        return copy(
            accepted = allOptional,
            rejected = emptySet(),
            consentedAtMs = System.currentTimeMillis(),
        )
    }

    /**
     * Returns a new state with all non-necessary categories rejected.
     */
    fun rejectingAll(): ConsentState {
        val allOptional = ConsentCategory.entries.filter { it.requiresConsent }.toSet()
        return copy(
            accepted = emptySet(),
            rejected = allOptional,
            consentedAtMs = System.currentTimeMillis(),
        )
    }

    /**
     * Returns a new state accepting only the specified [categories] (and rejecting all others).
     *
     * Passing [ConsentCategory.NECESSARY] in the set is a no-op — it will not appear in
     * either [accepted] or [rejected].
     */
    fun accepting(categories: Set<ConsentCategory>): ConsentState {
        val allOptional = ConsentCategory.entries.filter { it.requiresConsent }.toSet()
        val toAccept = categories.filter { it.requiresConsent }.toSet()
        val toReject = allOptional - toAccept
        return copy(
            accepted = toAccept,
            rejected = toReject,
            consentedAtMs = System.currentTimeMillis(),
        )
    }
}
